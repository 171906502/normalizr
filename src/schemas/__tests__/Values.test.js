/* eslint-env jest */
import { normalize, schema } from '../../';

describe(schema.Values.name, () => {
  it('normalizes the values of an object with the given schema', () => {
    const cat = new schema.Entity('cats');
    const dog = new schema.Entity('dogs');
    const valuesSchema = new schema.Values({
      dogs: dog,
      cats: cat
    }, (entity, key) => entity.type);

    expect(normalize({
      fido: { id: 1, type: 'dogs' },
      fluffy: { id: 1, type: 'cats' }
    }, valuesSchema)).toMatchSnapshot();
  });

  it('can use a function to determine the schema when normalizing', () => {
    const cat = new schema.Entity('cats');
    const dog = new schema.Entity('dogs');
    const valuesSchema = new schema.Values({
      dogs: dog,
      cats: cat
    }, (entity, key) => `${entity.type}s`);

    expect(normalize({
      fido: { id: 1, type: 'dog' },
      fluffy: { id: 1, type: 'cat' }
    }, valuesSchema)).toMatchSnapshot();
  });

  it('throws if cannot find a matching schema', () => {
    const dog = new schema.Entity('dogs');
    const valuesSchema = new schema.Values({
      dogs: dog
    }, (entity) => `${entity.type}s`);

    expect(() => normalize({ fluffy: { id: 1, type: 'cat' } }, valuesSchema)).toThrow();
  });

  it('filters out null and undefined values', () => {
    const cat = new schema.Entity('cats');
    const dog = new schema.Entity('dogs');
    const valuesSchema = new schema.Values({
      dogs: dog,
      cats: cat
    }, (entity, key) => entity.type);

    expect(normalize({
      fido: undefined,
      milo: null,
      fluffy: { id: 1, type: 'cats' }
    }, valuesSchema)).toMatchSnapshot();
  });
});
