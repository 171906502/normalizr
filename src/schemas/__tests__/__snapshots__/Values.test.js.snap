exports[`ValuesSchema can use a function to determine the schema when normalizing 1`] = `
Object {
  "entities": Object {
    "cats": Object {
      "1": Object {
        "id": 1,
        "type": "cat",
      },
    },
    "dogs": Object {
      "1": Object {
        "id": 1,
        "type": "dog",
      },
    },
  },
  "result": Object {
    "fido": 1,
    "fluffy": 1,
  },
}
`;

exports[`ValuesSchema normalizes the values of an object with the given schema 1`] = `
Object {
  "entities": Object {
    "cats": Object {
      "1": Object {
        "id": 1,
        "type": "cats",
      },
    },
    "dogs": Object {
      "1": Object {
        "id": 1,
        "type": "dogs",
      },
    },
  },
  "result": Object {
    "fido": 1,
    "fluffy": 1,
  },
}
`;
